README.md
# pygifme

![pug](http://f.cl.ly/items/0T0f2w2C2z3T343w0u37/pug.gif)

[![Travis status build](https://secure.travis-ci.org/shakaran/pygifme.png?branch=master)](http://travis-ci.org/shakaran/pygifme)

## Fucking animations. You need them.

pygifme is a simple command line tool to generate animated GIFs. It is a python port from the original ruby script gifme created by [@holman](https://twitter.com/holman)

## Installation

Install pygifme:

    pip3 install pygifme

You also can install via setup.py file:

    python3 setup.py install

You'll also need to install ImageMagick. 

On GNU/LINUX Debian systems:

    sudo apt-get install imagemagick

On OS X, this is easy using Homebrew:

    brew install imagemagick

This script is for Python >= 3.0. It also needs python module argcomplete if
you wish pygifme autocompletion in command line.

You can install with:

    sudo pip3 install argcomplete

Or via requirements.txt file:

   sudo pip3 install -r requirements.txt --use-mirrors

## Usage

    pygifme ~/Desktop/1.png ~/Desktop/2.png
    You now have a handsome animation at ~/Desktop/animated.gif

You can also glob, of course:

    pygifme ~/Desktop/*.jpg
    You now have a handsome animation at ~/Desktop/animated.gif

## CloudApp

Once your animation is finished up, we'll try to upload it to
[CloudApp](http://www.getcloudapp.com). If you have 
[`cloudapp`](https://github.com/holman/dotfiles/blob/master/bin/cloudapp)
script installed, we'll use that, otherwise we'll just skip this whole step.

## Super Advanced Usage

On some gifs, it's nice to have a smooth loop, so when it ends there's not a
jarring leap from the last frame to the first frame again. Use the `--reverse`
switch to create the animation like normal, and then reverse the frames and add
them to the animation so it looks like one smooth motion and back again.

    pygifme FILES --reverse

If you pass in a URL of a gif instead of FILES, we'll download that gif, split
it into its constituent frames, and let you recreate it. For example, you could
take a gif you find online and give it that `--reverse` look:

    pygifme http://tumblr.com/some-crazy.gif --reverse

You can also resize shit. We default to 500 pixels, but do whatever the fuck
you want.

    pygifme http://tumblr.com/some-crazy.gif --width=1000

For other options, check out the help:

    pygifme -h

## History

If you're curious, gifme was initially a few-line shell script in [Zach Holman
dotfiles](https://github.com/holman/dotfiles). Eventually it became clear that
animation is a fundamental part of our society, and I split it out into its own
tiny project.

If you're curious, Zach Holman featured the older gifme version [in a
screencast](http://zachholman.com/2011/01/automating-inefficiencies/) Zach Holman made
that describes how animated gifs are an integral part of working at GitHub.

I made this python port just for fun and practice ruby to python porting.

## Running unit tests

This is a simple script, but it can have unexpect behaviour doing weird things
with ImageMagick backend. So, if you want ensure the common behaviour, then you
can run the unit testing battery with:

    python3 -m unittest test

Also from setup.py (recommended):

    sudo python3 setup.py test

## Much Love

From [@holman](https://twitter.com/holman). Zach Holman loves you.

## About

Mantainer [@shakaran87](https://twitter.com/shakaran87).


CHANGELOG.rst
Version 0.1
------------

-   Create a port in python of ``gifme`` called ``pygifme``.


test.py
# -*- coding: utf-8; tab-width: 4; mode: python -*-
# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: t -*-
# vi: set ft=python sts=4 ts=4 sw=4 noet 

#    pygifme is a simple command line tool to generate animated GIFs. 
#    It is a python port from the original ruby script gifme created by
#    Zach Holman

#    Copyright (C) 2013 by Ángel Guzmán Maeso, shakaran at gmail dot com

#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 3 of the License, or
#    (at your option) any later version.

#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#    GNU General Public License for more details.

#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
import sys
from os import system, environ, EX_OK
from subprocess import check_output, CalledProcessError, STDOUT
import unittest

if sys.version_info[:3] < (3, 0, 0):
    print('Python 3.0 or higher is required, {0} found. Aborting.'.format('.'.join(map(str, sys.version_info[:3]))))
    sys.exit(1)

def disable_cloud_app():
    environ['DISABLE_CLOUPAPP'] = '1'
    
def enable_cloud_app():
    if environ.get('DISABLE_CLOUPAPP'):
        del environ['DISABLE_CLOUPAPP']

class TestBinaries(unittest.TestCase):
    """ Test Binaries """
        
    @unittest.skipIf(sys.version_info < (2, 6),
                     'argparse not supported in this python version')
    def test_argpase(self):
        """ Checks the presence of argparse module """
        try:
            import argparse
            assert True
        except ImportError as e:
            self.fail('The module argparse is required')
       
    def test_argcomplete(self):
        """ Checks the presence of argcomplete module """
        try:
            import argcomplete
            assert True
        except ImportError as e:
            self.fail('The module argcomplete is required. Try: sudo pip install argcomplete')
                 
    def test_binary(self):
        """ Checks the binary presence of pygifme """
        assert system('which pygifme 2>&1 > /dev/null') == 0
        
    def test_imagemagick(self):
        """ Checks the binary presence of imagemagick """
        assert system('which convert 2>&1 > /dev/null') == 0
        
    def test_cloudapp(self):
        """ Checks the binary presence of cloudapp """
        # Detect presence of cloudapp (no required at all)
        # gem install cloudapp
        if system('which cloudapp 2>&1 > /dev/null') == 0:
            self.assertTrue(True, 'cloudapp binary detected')
        else:
            self.skipTest('cloudapp no detected. Skipping because not mandatory')

class TestArguments(unittest.TestCase):
    """ Test Arguments """
    
    REMOTE_GIF = 'https://a248.e.akamai.net/camo.github.com/0b546aee911bb34aeaa3109081ec70d15b159d2c/687474703a2f2f662e636c2e6c792f6974656d732f3054306632773243327a335433343377307533372f7075672e676966'
    
    def setUp(self):
        disable_cloud_app()
    
    def tearDown(self):
        enable_cloud_app()
             
    def test_no_files(self):
        """ Test invocation without FILES argument """
        
        try:
            check_output(['pygifme'], stderr = STDOUT)
            self.fail('error: pygifme must run with FILES argument')
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            result = 'the following arguments are required: FILES'
            self.assertIn(result, str(e.output))
    
    def test_unexistant_file(self):
        """ Test invocation with unexistant FILES argument """
        try:
            output = check_output(['pygifme', 'some_unexistant_file'], stderr = STDOUT)
            self.fail('Failed because the file should be unexistant:' + str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            
            result = 'does not exist'
            self.assertIn(result, str(e.output))
            
        
    def test_one_file(self):
        """ Test invocation with one FILES argument """
        output = check_output(['pygifme', 'test/1.jpg'], stderr = STDOUT)
        
        result = 'You now have a handsome animation at'
        self.assertIn(result, str(output))
    
    def test_two_files(self):
        """ Test invocation with two FILES argument """
        output = check_output(['pygifme', 'test/1.jpg', 'test/2.jpg'], stderr = STDOUT)
                              
        result = 'You now have a handsome animation at'
        self.assertIn(result, str(output))

    def test_global_files(self):
        """ Test invocation with global FILES argument """
        try:
            output = check_output(['pygifme', 'test/*.jpg'], stderr = STDOUT)
            
            result = 'You now have a handsome animation at'
            self.assertIn(result, str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            self.fail('Failed because:' + str(e.output))
    
    def test_remote_file(self):
        """ Test invocation with remote FILES argument """
        try:
            output = check_output(['pygifme', self.REMOTE_GIF], stderr = STDOUT)
                                  
            result = 'You now have a handsome animation at'
            self.assertIn(result, str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            self.fail('Failed because:' + str(e.output))
    
    def test_missing_delay(self):
        """ Test invocation with missing DELAY argument """
        try:
            output = check_output(['pygifme', '-d'], stderr = STDOUT)
            self.fail('Failed because the file should be unexistant:' + str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            
            result = 'expected one argument'
            self.assertIn(result, str(e.output))
        
        try:
            output = check_output(['pygifme', '-d', 'test/1.jpg'], stderr = STDOUT)
            self.fail('Failed because the file should be unexistant:' + str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            
            result = 'invalid int value'
            self.assertIn(result, str(e.output))
    
    def test_delay(self):
        """ Test invocation with DELAY argument """
        try:
            output = check_output(['pygifme', '-d', '25', 'test/1.jpg'], stderr = STDOUT)
            
            result = 'You now have a handsome animation at'
            self.assertIn(result, str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            self.fail('Failed because:' + str(e.output))
    
    def test_missing_width(self):
        """ Test invocation with missing WIDTH argument """
        try:
            output = check_output(['pygifme', '-w'], stderr = STDOUT)
            self.fail('Failed because the file should be unexistant:' + str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            
            result = 'expected one argument'
            self.assertIn(result, str(e.output))
        
        try:
            output = check_output(['pygifme', '-w', 'test/1.jpg'], stderr = STDOUT)
            self.fail('Failed because the file should be unexistant:' + str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            
            result = 'invalid int value'
            self.assertIn(result, str(e.output))
    
    def test_with(self):
        """ Test invocation with WITH argument """
        try:
            output = check_output(['pygifme', '-w', '200', 'test/1.jpg'], stderr = STDOUT)
            
            result = 'You now have a handsome animation at'
            self.assertIn(result, str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            self.fail('Failed because:' + str(e.output))
            
    def test_reverse(self):
        """ Test invocation with -r/--reverse argument """
        try:
            output = check_output(['pygifme', '-r', 'test/1.jpg', 'test/2.jpg'], stderr = STDOUT)
            
            result = 'You now have a handsome animation at'
            self.assertIn(result, str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            self.fail('Failed because:' + str(e.output))
    
    def test_ouput(self):
        """ Test invocation with -o/--output argument """
        try:
            output = check_output(['pygifme', '-o', '/tmp/pygifme', 'test/1.jpg', 'test/2.jpg'], stderr = STDOUT)
            
            result = 'You now have a handsome animation at'
            self.assertIn(result, str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            self.fail('Failed because:' + str(e.output))
    
    def test_quiet(self):
        """ Test invocation with -q/--quiet argument """
        try:
            output = check_output(['pygifme', '-q', 'test/1.jpg', 'test/2.jpg'], stderr = STDOUT)
            
            result = ''
            self.assertIn(result, str(output))
        except CalledProcessError as e:
            self.assertNotEqual(e.returncode, EX_OK)
            self.fail('Failed because:' + str(e.output))
    
    def test_help(self):
        """ Test invocation with -h/--help argument """
        output = check_output(['pygifme', '-h'], stderr = STDOUT)
        
        result = 'usage: pygifme'
        self.assertIn(result, str(output))
        
        output = check_output(['pygifme', '--help'], stderr = STDOUT)
        
        self.assertIn(result, str(output))


__init__.py
# -*- coding: utf-8; tab-width: 4; mode: python -*-
# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: t -*
# vi: set ft=python sts=4 ts=4 sw=4 noet

#    pygifme is a simple command line tool to generate animated GIFs.
#    It is a python port from the original ruby script gifme created by
#    Zach Holman

#    Copyright (C) 2013 by Ángel Guzmán Maeso, shakaran at gmail dot com

#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 3 of the License, or
#    (at your option) any later version.

#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#    GNU General Public License for more details.

#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

__package__ = 'pygifme'
__version__ = '0.1'
__all__ = ['.pygifme']

def main(): # For setuptools entry point
    from . import pygifme
    pygifme.main()

pygifme.py
#!/usr/bin/env python3
# -*- coding: utf-8; tab-width: 4; mode: python -*-
# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: t -*
# vi: set ft=python sts=4 ts=4 sw=4 noet

# pylint: disable-msg=R0903,C0103,C0301

#    pygifme is a simple command line tool to generate animated GIFs.
#    It is a python port from the original ruby script gifme created by
#    Zach Holman

#    Copyright (C) 2013 by Ángel Guzmán Maeso, shakaran at gmail dot com

#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 3 of the License, or
#    (at your option) any later version.

#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#    GNU General Public License for more details.

#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

# PYTHON_ARGCOMPLETE_OK

def main():
    from os import system, mkdir, listdir, environ, path, chdir, getcwd, \
                   EX_USAGE, EX_DATAERR, EX_CONFIG, EX_CANTCREAT
    from time import gmtime, strftime
    from sys import exit, path as spath
    from argparse import ArgumentParser, ONE_OR_MORE, ArgumentDefaultsHelpFormatter
    from argcomplete import autocomplete
    
    # Avoid problems with /usr/local/bin first in sys.path
    if path.dirname(__file__) == spath[0]:
        spath.append(spath[0])
        del spath[0]
    
    try:
        from pygifme import __version__
    except ImportError as e:
        #from . import __version__
        __version__ = '0.1' # @Fixme: Bad package import from /usr/local/bin/pygifme.py
        
    description = '  FILES can be listed out, like `file1.jpg file2.jpg`, or it\n' \
                  '  can be a normal shell glob, like `*.jpg`.'
            
    parser = ArgumentParser(
                             prog                  = 'pygifme',
                             description           = description,
                             epilog                = None,
                             parents               = [],
                             formatter_class       = ArgumentDefaultsHelpFormatter,
                             prefix_chars          = '-',
                             fromfile_prefix_chars = None,
                             argument_default      = None,
                             conflict_handler      = 'error',
                             add_help              = True,
                            )
    
    def valid_directory(directory):
        if path.exists(directory):
            if path.isdir(directory):
                return directory
            else:
                parser.error('Path {0} is not a directory'.format(directory))
                return directory
        else:
            parser.error('Directory path {0} does not exist'.format(directory))
                       
    parser.add_argument('-r', '--reverse',
                        action         = 'store_true',
                        dest           = 'reverse',
                        default        = False,
                        help           = 'Reverse the GIF to make it loopable')
    
    parser.add_argument('-o', '--output',
                        action         = 'store',
                        metavar        = '/path/to/output',
                        type           = lambda d:valid_directory(d),
                        dest           = 'output',
                        choices       = None,
                        help           = 'Set the animation\'s output directory')
    
    parser.add_argument('-d', '--delay',
                        action         = 'store',
                        metavar        = 'DELAY',
                        dest           = 'delay',
                        default        = 20,
                        type           = int,
                        choices        = None,
                        help           = 'Set the delay between frames')
    
    parser.add_argument('-w', '--width',
                        action         = 'store',
                        metavar        = 'PIXELS',
                        dest           = 'width',
                        default        = 500,
                        type           = int,
                        choices        = None,
                        help           = 'Set the width of the image')
    
    parser.add_argument('-q', '--quiet',
                        action         = 'store_true',
                        dest           = 'quiet', 
                        default        = False,
                        help           = 'Don\'t print status messages to stdout')
    
    parser.add_argument(option_strings = ['FILES'], 
                        metavar        = 'FILES', 
                        nargs          = ONE_OR_MORE,
                        type=str,
                        dest           = 'FILES',
                        help           = 'One or more files to process')
    
    parser.add_argument('-v', '--version', 
                        action  ='version',
                        version ='%(prog)s {version}'.format(version = __version__),
                        help    = 'Shows the program version')
    
    autocomplete(parser)
    args = parser.parse_args()
    
    #print(vars(args)) # For debugging
    
    if system("which convert 2>&1 > /dev/null") != 0:
        parser.error('You need to install ImageMagick first.\n\n' \
               'If you\'re on GNU/LINUX Debian systems, this should be as easy as:\n'\
               '  sudo apt-get install imagemagick\n' \
               'If you\'re on a Mac, this should be as easy as:\n' \
               '  brew install imagemagick')
        exit(EX_CONFIG)
        
    if not args.FILES: # no files given
        parser.error('no FILES given to process')
        exit(EX_USAGE)
    else:
        for pfile in args.FILES:
            if pfile[0:4] != 'http': # skip remote files
                if path.exists(pfile):
                    if not path.isfile(pfile):
                        parser.error('{0} is not a valid file'.format(pfile))
                else:
                    import glob
                    result_glob = glob.glob(pfile)
                    if not result_glob:
                        parser.error('File {0} does not exist'.format(pfile))
                    else:
                        for gfile in result_glob:
                            if path.exists(gfile):
                                if not path.isfile(gfile):
                                    parser.error('{0} is not a valid file'.format(gfile))
                            else:
                                parser.error('File {0} does not exist'.format(gfile))
        
    # WORKING WITH REMOTE FILES
    if args.FILES[0][0:4] == 'http':
        from urllib import request
        from shutil import rmtree
        
        if path.exists('/tmp/pygifme'):
            rmtree('/tmp/pygifme')
            
        mkdir('/tmp/pygifme', 0o777)
        
        local_path = '/tmp/downloaded-pygifme.gif'
    
        with open(local_path, mode='wb') as wfile:
            remote_file = request.urlopen(args.FILES[0]).read()
            wfile.write(remote_file)
    
        wd = getcwd()
        chdir('/tmp/pygifme')
        status = system('convert {0} -coalesce frame_%03d.gif'.format(local_path))
        chdir(wd)
        
        if status != 0:
            parser.error('Could not process remote file {0}'.format(local_path))
            exit(EX_DATAERR)
        
        del args.FILES[0]
        args.FILES += [path.join('/tmp/pygifme/', f) for f in listdir('/tmp/pygifme/')]
    
    if args.reverse:
        args.FILES += args.FILES[1:-2]
    
    if not args.output:
        import subprocess
        desktop = str(subprocess.check_output(['xdg-user-dir', 'DESKTOP'], universal_newlines = True)).strip()
        
        if not desktop:
            home = environ.get('HOME')
            if not home:
                home = '/tmp/pygifme'
                mkdir(home, 0o777)
            else:
                home = '{0}/Desktop'.format(home)
        else:
            home = desktop

        args.output = home
    
    args.output = str(args.output) + '/animated-{0}.gif'.format(strftime('%F_%Hh-%Mm-%Ss', gmtime()))
               
    cmd = 'convert -delay {0} -loop 0 -resize {1} -layers OptimizeTransparency {2} {3}'.format(int(args.delay), int(args.width), ' '.join(args.FILES), str(args.output))
    if system(cmd) == 0:
        if not args.quiet:
            print('You now have a handsome animation at {0}'.format(args.output))
    else:
        parser.error('Something broke when we were animating your gif. Shit.')
        exit(EX_CANTCREAT)
    
    if system('which cloudapp 2>&1 > /dev/null') == 0 and not environ.get('DISABLE_CLOUPAPP'):
        if not args.quiet:
            print('Now we\'re uploading it to CloudApp')
        system('cloudapp {0}'.format(args.output))

if __name__ == '__main__':
    main()